src/main/java/io/github/oliviercailloux/j_voting/preferences/classes/CompletePreferenceImpl.java

```diff
@@ -0,0 +1,128 @@
+     */
+    public static CompletePreferenceImpl createCompletePreferenceImpl(
+                    ImmutableList<ImmutableSet<Alternative>> preference,
+                    Voter voter) {
+        LOGGER.debug("Factory CompletePreferenceImpl");
+        Preconditions.checkNotNull(preference);
+        Preconditions.checkNotNull(voter);
+        return new CompletePreferenceImpl(preference, voter);
+    }
+
+    /**
+     * 
+     * @param preference : <code> not null </code>
+     * @param voter      : <code> not null </code>
+     */
+    private CompletePreferenceImpl(
+                    ImmutableList<ImmutableSet<Alternative>> preference,
+                    Voter voter) {
+        LOGGER.debug("Constructor CompletePreferenceImpl");
+        this.voter = voter;
+        this.preference = preference;
+        this.graph = createGraph(preference);
+    }
+
+    private ImmutableGraph<Alternative> createGraph(
+                    ImmutableList<ImmutableSet<Alternative>> preference) {
+        MutableGraph<Alternative> newGraph = GraphBuilder.directed()
+                        .allowsSelfLoops(true).build();
+        Alternative lastSetLinker = null;
+        for (ImmutableSet<Alternative> set : preference) {
+            Alternative rememberAlternative = null;
+            for (Alternative alternative : set) {
+                if (!Objects.isNull(lastSetLinker)) {
+                    newGraph.putEdge(lastSetLinker, alternative);
+                    lastSetLinker = null;
+                }
+                newGraph.putEdge(alternative, alternative);
+                if (!Objects.isNull(rememberAlternative)) {
+                    newGraph.putEdge(rememberAlternative, alternative);
+                    newGraph.putEdge(alternative, rememberAlternative);
+                }
+                rememberAlternative = alternative;
+            }
+            lastSetLinker = rememberAlternative;
+        }
+        ;
+        System.out.println(Graphs.transitiveClosure(newGraph).toString());
+        return ImmutableGraph.copyOf(newGraph);
+    }
+
+    @Override
+    public ImmutableSet<Alternative> getAlternatives() {
+        Set<Alternative> returnedSet = new HashSet<>();
+        for (ImmutableSet<Alternative> set : preference) {
+            for (Alternative alternative : set) {
+                returnedSet.add(alternative);
+            }
+        }
+        return ImmutableSet.copyOf(returnedSet);
+    }
+
+    @Override
+    public Voter getVoter() {
+        return this.voter;
+    }
+
+    @Override
+    public int getRank(Alternative a) {
+        Preconditions.checkNotNull(a);
+        for (ImmutableSet<Alternative> set : preference) {
+            if (set.contains(a))
+                return preference.indexOf(set) + 1;
+        }
+        throw new NotFoundException("Alternative not found");
+    }
+
+    @Override
+    public ImmutableSet<Alternative> getAlternatives(int rank) {
+        Preconditions.checkArgument(rank > 0);
```

Comment:
Asymétrie bizarre entre rang trop petit et rang trop grand (non vérifié explicitement). Mieux vaut laisser l’exception par défaut à mon avis (out of range).